### ISSUE TITLE
Handle multi image annotation

### ISSUE THREAD

**Issue by rsimon** [recogito/annotorious-openseadragon#32]
We need to figure out what's needed to support annotation of a [multi-tilesource setup](https://openseadragon.github.io/examples/multi-image/). The key question to answer is how to get proper image coordinates.

**Comment by rsimon** [comment]
A workspace/playground repo for this is at:

https://github.com/IIIF-Commons/osd-canvasworld-annotorious/

**Comment by umesh-timalsina** [comment]
@rsimon Any progress on this front? I am currently trying to bulk annotate multiple tilesources loaded in a single viewport and getting the following error:
```js
Uncaught TypeError: Cannot read property 'naturalWidth' of undefined
    at oa (openseadragon-annotorious.min.js:2)
    at new e (openseadragon-annotorious.min.js:2)
    at i.startDrawing (openseadragon-annotorious.min.js:2)
    at i.start (openseadragon-annotorious.min.js:2)
    at $.MouseTracker.pressHandler (openseadragon-annotorious.min.js:2)
    at updatePointersDown (openseadragon.js:5930)
    at onPointerDown (openseadragon.js:5472)
    at SVGSVGElement.pointerdown (openseadragon.js:3340)
```

This is what my viewport looks like:

![image](https://user-images.githubusercontent.com/11476842/119690346-36e9a200-be0f-11eb-872f-313465fb1066.png)

**Comment by umesh-timalsina** [comment]
I will note here that the error message [above](https://github.com/recogito/annotorious-openseadragon/issues/32#issuecomment-848884768) seems to occur when you initialize `annotoriousOSD` without any tileSources to begin with. The images are added to the viewPort only after the plugin is in the example above.

**Comment by rsimon** [comment]
Unfortunately, there's no progress on this yet. I've been wanting to look into this. But frankly, it kept dropping down the priority list, since none of my day job projects currently needs it.

I think it might be a bigger task. Right now, Annotorious implicitly assumes that the OSD world coordinate space equals the image coordinate space; and the dimensions are initialized when the Annotorious instance is created. (Which is exactly in line with the error messages you are seeing.)

To address this, we'd first need to figure out: the right way to translate between world coordinates and image coordinates (and vice versa).

Then, whenever an annotation is created, we'd need to find a way to figure out which tilesource it belongs to (there should be an OSD API method for that, I guess).

The rest is plumbing. It's probably not entirely trivial either. But once there's a first working hack, at least we'll be able to see what it takes to build this.

As I said: I'd be really interested too see (and help with!) this. But my gut feeling is that exceeds my weekend side project time budget at the moment.

**Comment by umesh-timalsina** [comment]
@rsimon Thanks for the reply. 
I am willing to add this feature. Few suggestions/questions in the current API?
> I think it might be a bigger task. Right now, Annotorious implicitly assumes that the OSD world coordinate space equals the image coordinate space; and the dimensions are initialized when the Annotorious instance is created. (Which is exactly in line with the error messages you are seeing.)

Can we leverage the `OpenSeaDragon.viewer.world's` [`add-item`](https://openseadragon.github.io/docs/OpenSeadragon.World.html#.event:add-item) event for getting the dimensions? I tried it locally with my fork for the demo in this repository. Everything seem to work fine for a single image. Admittedly, I haven't looked into `DrawingTools` code to know how it works. Is it the case that it requires a single image source url for a single drawer instance?

> To address this, we'd first need to figure out the right way to translate between world coordinates and image coordinates (and vice versa). 

I found an few API points in OpenSeadragon `TiledImage` class:
 (a.) [viewPortToImageCoordinates](https://openseadragon.github.io/docs/OpenSeadragon.TiledImage.html#viewportToImageCoordinates) 

(b.) [ImageToViewPortCordinates](https://openseadragon.github.io/docs/OpenSeadragon.TiledImage.html#imageToViewportCoordinates)

> Then, whenever an annotation is created, we'd need to find a way to figure out which tilesource it belongs to (there should be an OSD API method for that, I guess).

I have experimented with a way to figure this out. But it might not be optimal, On any mouse event, you can detect  which tiled image source's bounding box the coordinates belong to.

> The rest is plumbing. It's probably not entirely trivial either. But once there's a first working hack, at least we'll be able to see what it takes to build this.

Agreed, but I would like to know what details would be necessary, I am fairly constrained in terms of time, albeit we can cross the bridge when we get there :). Thanks.

**Comment by rsimon** [comment]
- Yes, the `add-overlay` event for getting dimensions sounds right. 
- `viewportToImageCoordinates` (and vice versa) doesn't work in multi-image mode (see [docs here](https://openseadragon.github.io/docs/OpenSeadragon.Viewport.html#viewportToImageCoordinates)). But there are similar methods on the `TiledImage` sources. Once you've figured out which tilesource the annotation is on, you should be able to use those.

The main problem for the plumbing is this: AnnotoriousOSD re-uses the drawing tools from "standard" Annotorious. Therefore, the tools need some kind of common abstraction for the underlying image. That's simple as long as both are only dealing with a single image. (The tools really just expect an object that has `naturalWidth` and `naturalHeight` properties.) But things get much trickier with OSD multi-image mode.

Changes are (at least) necessary:

1. [here](https://github.com/recogito/annotorious-openseadragon/blob/main/src/OSDAnnotationLayer.js#L44-L49), where width and height are stored, so that the tools can later access them; and
2. [here](https://github.com/recogito/annotorious/blob/main/src/tools/Tool.js#L25-L38) in the base class for tool implementations, where mouse pointer coordinates are translated to image coordinates

For the first part, I think it's sufficient to store the size of the OSD world (and re-size it when a new tilesource is added). The second part is the tricky bit & needs experimentation. AnnotoriousOSD will sometimes need to convert between mouse and world coordinates (to draw the shapes); but sometimes between mouse (or world) and tilesource image coordinates, to store and render shapes. Wouldn't know off the top of my head what's needed where exactly in the code. Figuring that out might be a bit of painful step-by-step process. But any help you can give would be massively appreciated!

**Comment by umesh-timalsina** [comment]
Further update on this:
1. There are two modes currently used in OSD that might be of concern for this feature support:
  ## Sequence Mode
 For sequence mode, there are multiple tile sources but the `OSDViewer` opens only one image at a time. Currently, with the calculations made for drawing annotations work fine and the [`open`](https://github.com/recogito/annotorious-openseadragon/blob/2c16cbf72cb9644903ef76993ef3bf309b48b726/src/OSDAnnotationLayer.js#L40) event is fired for every page change in sequence mode. However, the currently the annotation layer doesn't take into account the underlying source for which the annotation is loaded. So, if you annotate an image and switch pages in the viewer, the annotations for previous image are displayed as well (Example below): 
 
Page 1
![image](https://user-images.githubusercontent.com/11476842/121961235-c4445600-cd2c-11eb-9043-2375d5739fce.png)

Page 2 
![image](https://user-images.githubusercontent.com/11476842/121961655-4fbde700-cd2d-11eb-824c-9f0039b7da6f.png)

Since the base calculations for annotating the image works and events are fired correctly, we need the following additions to support it in My Opinion:
1. On page change of the viewer, partition the list of annotations into displayed vs hidden based on `target.source` url for the annotation i.e. if the url of the current tile source in the OSDViewer matches with the `target.source`, it goes to displayed pool vs hidden pool for the opposite case.
2. Handle all the events like `cancelSelection` if there's an active selection in the viewPort.

This approach works only for sequence mode. I have a reasonable solution for the partitioning approach described above based on query selectors but I didn't find any other way to hide an annotation based on my understanding of the `annotorious`'s public API. 

 ## Collection Mode
Still investigating the collection mode, but I did fiddle around with the following [`pen`](https://codepen.io/iangilman/pen/PqxvgN) that does the hit detection in `collection` mode. Based on the code in the coordinate translation functions mentioned [above](https://github.com/recogito/annotorious-openseadragon/issues/32#issuecomment-849357113), if we can find a way to map the mouse coordinates to the image coordinates, I think we have a decent starting point.

**Comment by rsimon** [comment]
Thanks so much for digging into this! I'll can probably look into this more closely on the weekend. But meanwhile I also wanted to point to issue #52, which is specifically about sequence mode. 

My take on this is that Annotorious shouldn't do too much in this case, and leave replacing of the annotations up to the host application. (Annotorious doesn't check whether the URI of the image and the `source` attribute match up. And there has been discussion about this causing potential pitfalls.) I.e. the core probably should just make sure the image dimensions get updated accordingly, and the annotation layer gets wiped.

As for the common functionality of paginating the annotations and applying the right subset: perhaps that's something better handled in a (very) small "sequenceMode plugin". This way, users of sequence mode wouldn't need to re-write the same pagination code themselves, while the average user (who isn't using sequence mode) wouldn't need to deal with nasty surprises if image URL and their annotations' `source` attribute don't match up.

**Comment by umesh-timalsina** [comment]
Further thoughts on this, its not clear that how multiple images should be handled by the underlying annotation layer, If there are multiple images in the ViewPort, one idea is to create multiple annotation layers; one for for each image. And tie the boundary of the annotation layer with the Image Bounding Box. Since, we can get [`Image-Coordinates`](https://umesh-timalsina.github.io/osd-coordinates/) from the viewport coordinates, Further thoughts on this might be necessary, but I think this will be a good way to handle individual images in the viewport. This way any annotations drawn on boundaries will be taken care of automatically.

**Comment by rsimon** [comment]
This sounds like a good approach to managing multiple images. I think this should also allow the re-use of the drawing tools without modification. 

Extensions to the Annotator will probably have to be substantial. But it's definitely worth a try.

**Comment by umesh-timalsina** [comment]
For anybody willing to try this feature, a super primitive version has been implemented in [this fork](https://github.com/symbench/annotorious-openseadragon). Upon further reflection on this issue, this feature can also be added as a plugin in my opinion (like the sequence mode plugin). If we are able to hookup a custom annotation layer to the app, that would work as well, however keeping two annotation layers in sync would be a maintenance. I don't know what the plans for this project are, but supporting multi-image annotation has become a business requirement for one of our project. 

Regardless, the biggest roadblock for me is tying an individual image in the view-port to a particular annotation layer. This is absolutely necessary for `setAnnotations` method to discover which image is loaded in which tile source. The branch above takes a URL based approach. There might be other minor patches necessary, but the functionality exists.

**Comment by rsimon** [comment]
@umesh-timalsina sorry I didn't look into your fork more deeply yet. But this seems to work pretty well already!

My initial feeling would have been that separating this into two annotation layers (=two separate SVG overlays) could make this easier. Or at least allow to break the code up a bit (into the actual layer, and then some kind of managing class "above" in the hierarchy.) That _could_ make it easier to restrict drawing to a specific image, since you could set the size of the SVG the same as that of the image. (And then maybe even highlight the image region/border as the user moves the mouse over it, so people know which image they are on.)

But perhaps you are right that the need to keep things in sync raises lots of other issues. Do you mind sending a pull request for this? (We could, e.g. pull it into a dedicated branch while the architecture might still be changing.)

**Comment by rsimon** [comment]
On second thought, constraining already works pretty well already at the level of the `<g>` elements, with the image size set through naturalWidth/Height. My feeling is that somehow splitting the `OSDAnnotationLayer` into two parts might be useful and aid readability. (And maybe provide a clue whether this really _could_ be solved in a plugin-based way, too.) 

But IMO this works great, and it looks like it's really only about a bit of refactoring more than anything else!

**Comment by umesh-timalsina** [comment]
@rsimon Could you elaborate more on splitting `OSDAnnotationLayer` into two parts?

**Comment by rsimon** [comment]
Thanks for the PR!

I think what might work is to move all the code that relates to a single TiledImage into a separate class. Maybe something called "AnnotatableImage" perhaps. That would hold the ref to the SVG group, the tool & event handlers, methods to encapsulate the queries "is this point on this image", "is this annotation on this image" etc. And then the OSDAnootationLayer would hold a list of AnnotatableImages, and handle the general management - deferred annotation loading, resizing the SVG layer etc.

**Comment by umesh-timalsina** [comment]
@rsimon Thanks for keeping the feature branch here. I will try to incorporate what you suggested, it makes sense to have an `AnnotableImage` class instead of iterating through a list every time.
